src/ch/ch_process.c

```diff
@@ -436,6 +438,80 @@ static int virCHProcessSetupThreads(virDomainObjPtr vm)
     return ret;
 }
 
+/*
+ * Procfs Thread watcher
+ * This thread is started when a Domain boots up and
+ * exists as long as the domain is active. The duty
+ * of the thread is to watch for any tid changes for
+ * the VM and then update the cgroups accordingly.
+ * Threads can change in the case of following events:
+ * - VM reboot
+ * - Device activation
+ *
+ * TODO: Parsing procfs frequently is ugly and not scalable
+ *       Better option would be to have cloud-hypervisor
+ *       return the thread information via API query or
+ *       probably to have a monitor interface like what
+ *       qemu has.
+ */
+
+/*
+ * Hard coding the polling interval for the watcher thread
+ * to be 1 second.
+ */
+#define THREAD_WATCHER_POLL_INTERVAL    G_USEC_PER_SEC
+
+static void chProcessWatchThreads(void *data)
+{
+    virDomainObjPtr vm = (virDomainObjPtr)data;
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+
+    VIR_DEBUG("VM Thread watcher starting");
+    while (g_atomic_int_get(&priv->vmThreadWatcherStop) == 0) {
+
+        if (virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) < 0) {
```

Comment:
So I'm looking around at our code base a bit and I'm noticing that we don't have a lock around the vm and I'm worried that we've been getting away with it so far because we aren't using threads but this is going to start going poorly now. The comment above `virCHDomainObjBeginJob` calls out needing to have the obj locked which I think is the vm in this case before the function can be called.

Reply:
Thanks for pointing this out. I have implemented locking explicitly now and should be okay. I went through couple other places and looks like we do lock vm implicitly by the functions like `virDomainObjListAdd`  and `chDomObjFromDomain` etc. And we implicitly unlock by calling `virDomainObjEndAPI`. I haven't verified all the places yet. But I think we should be okay.

I shall do a thorough scan later and see if we missed locking in any places and will fix it as a separate effort!